[HUDI-5341] CleanPlanner retains earliest commits must not be later than earliest pending commit

SteNicholas opened the pull request:
### Change Logs

`CleanPlanner` should retain the earliest commits must not be later than the earliest pending commit. Meanwhile, `HoodieTimelineArchiver` should retain the clustering commit which instant can not be archived unless we ensure that the replaced files have been cleaned, without the replaced files metadata on the timeline, the fs view would expose duplicates for readers.

### Impact

`CleanPlanner` retains the earliest commits must not be later than the earliest pending commit and `HoodieTimelineArchiver` retains the clustering commit which instant can not be archived unless we ensure that the replaced files have been cleaned

### Risk level (write none, low medium or high below)

_If medium or high, explain what verification was done to mitigate the risks._

### Documentation Update

_Describe any necessary documentation update if there is any new feature, config, or user-facing change_

- _The config description must be updated if new configs are added or the default value of the configs are changed_
- _Any new feature or user-facing change requires updating the Hudi website. Please create a Jira ticket, attach the
  ticket number here and follow the [instruction](https://hudi.apache.org/contribute/developer-setup#website) to make
  changes to the website._

### Contributor's checklist

- [x] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [x] Change Logs and Impact were stated clearly
- [x] Adequate tests were added if applicable
- [x] CI passed

leesf commented:
cc @stream2000

SteNicholas commented:
@danny0405, @zhuanshenbsj1, I have created this pull request to fix the problem mentioned in #7405 in which the implementation is a little complex. PTAL.

danny0405 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/clean/CleanPlanner.java:
```diff
@@ -487,7 +487,24 @@ public Option<HoodieInstant> getEarliestCommitToRetain() {
+        // Earliest commit to retain must not be later than the earliest pending commit
+        earliestCommitToRetain =
+            commitTimeline.nthInstant(commitTimeline.countInstants() - commitsRetained).map(nthInstant -> {
+              if (nthInstant.compareTo(earliestPendingCommits.get()) <= 0) {
```
Why we need this check ? Can the cleaner cleans an instant that is not complete ?

SteNicholas replied:
@danny0405, this check is to ensure the earliest commit to retain must not be later than the earliest pending commit. The cleaner couldn't clean uncompleted instant and should keep the linear cleaning for incremental clean mode.

danny0405 replied:
Oh, i see, this change has effect for incremental cleaning, for normal cleaning, do we need this change ?

SteNicholas replied:
@danny0405, the `getPartitionPathsForFullCleaning` doesn't invoke this method.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/util/ClusteringUtils.java:
```diff
@@ -224,4 +225,36 @@ public static List<HoodieInstant> getPendingClusteringInstantTimes(HoodieTableMe
+        activeTimeline.getCleanerTimeline().filter(instant -> !instant.isCompleted()).firstInstant();
+    if (cleanInstantOpt.isPresent()) {
+      // The first clustering instant of which timestamp is greater than or equal to the earliest commit to retain of
+      // the clean metadata.
```
There is no need to do this check if there is no clustering instant on the timeline at all.

SteNicholas replied:
@danny0405, +1, I have added the check whether the completed replace timeline is empty.

zhuanshenbsj1 commented:
> @danny0405, @zhuanshenbsj1, I have created this pull request to fix the problem mentioned in #7405 in which the implementation is a little complex. PTAL.

LGTM. Clean clustering instant sequentially，to keep no clean holes in timeline like archive seems much more easily to implement.

danny0405 commented on hudi-common/src/main/java/org/apache/hudi/common/util/ClusteringUtils.java:
```diff
@@ -224,4 +225,37 @@ public static List<HoodieInstant> getPendingClusteringInstantTimes(HoodieTableMe
+    HoodieTimeline replaceTimeline = activeTimeline.getCompletedReplaceTimeline();
+    if (!replaceTimeline.empty()) {
+      Option<HoodieInstant> cleanInstantOpt =
+          activeTimeline.getCleanerTimeline().filter(instant -> !instant.isCompleted()).firstInstant();
```
+1 with this part, although this is a weak ensurence.

SteNicholas commented:
@leesf, @yihua, could you please review this pull request? I have addressed above comments from @danny0405.

leesf commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/HoodieTimelineArchiver.java:
```diff
@@ -432,6 +433,11 @@ private Stream<HoodieInstant> getCommitInstantsToArchive() {
 
+      // The clustering commit instant can not be archived unless we ensure that the replaced files have been cleaned,
+      // without the replaced files metadata on the timeline, the fs view would expose duplicates for readers.
+      Option<HoodieInstant> oldestInstantToRetainForClustering =
```
this name is a little confused with `oldestPendingCompactionAndReplaceInstant` below.

SteNicholas replied:
@leesf, refer to the naming of `oldestInstantToRetainForCompaction`.

SteNicholas commented:
@yihua, could you please review this pull request? @leesf has approved this changes.

yihua commented:
@SteNicholas Sorry for the delay.  I'll review the PR soon.

SteNicholas commented:
@yihua, any comments for this pull request?

yihua reviewed (approved):
The logic to determine the earliest instant to archive wrt clustering seems to be okay but it is very complex now.  It would be good to think about how to simplify this as a whole.

yihua commented:
[HUDI-5493](https://issues.apache.org/jira/browse/HUDI-5493) for revisiting the logic.

yihua commented:
@SteNicholas could you check the CI failure?

boneanxs commented on hudi-common/src/main/java/org/apache/hudi/common/util/ClusteringUtils.java:
```diff
@@ -224,4 +225,37 @@ public static List<HoodieInstant> getPendingClusteringInstantTimes(HoodieTableMe
+        // The first clustering instant of which timestamp is greater than or equal to the earliest commit to retain of
+        // the clean metadata.
+        HoodieInstant cleanInstant = cleanInstantOpt.get();
+        String earliestCommitToRetain =
```
`earliestCommitToRetain` might be null if we use `KEEP_LATEST_FILE_VERSIONS`, here we need to check it.

Also, if we use `KEEP_LATEST_FILE_VERSIONS`, not sure if this logic still is valid.

SteNicholas commented:
@yihua, I have already fixed the CI failure and updated the `ClusteringUtils`. PTAL.

SteNicholas commented:
@hudi-bot run azure

SteNicholas commented:
@yihua, I have already fixed the CI failure and updated the `ClusteringUtils`. Meanwhile the current failure has nothing to do with this change. PTAL.

SteNicholas commented:
@hudi-bot run azure

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/clean/CleanPlanner.java:
```diff
@@ -487,7 +487,24 @@ public Option<HoodieInstant> getEarliestCommitToRetain() {
+        // Earliest commit to retain must not be later than the earliest pending commit
+        earliestCommitToRetain =
+            commitTimeline.nthInstant(commitTimeline.countInstants() - commitsRetained).map(nthInstant -> {
+              if (nthInstant.compareTo(earliestPendingCommits.get()) <= 0) {
```
We should use the timestamp compare apis in HoodieTimeline 
```
HoodieTimeline.compareTimestamps(commit1Ts, GREATER_THAN, commit2Ts)
```
